skyportal/facility_apis/lt.py

```diff
@@ -0,0 +1,799 @@
+        )
+        contact = etree.SubElement(project, 'Contact')
+        etree.SubElement(contact, 'Username').text = altdata["username"]
+        etree.SubElement(contact, 'Name').text = altdata["username"]
+        etree.SubElement(contact, 'Communication')
+        cancel = etree.tostring(cancel_payload, encoding='unicode', pretty_print=True)
+
+        client = Client(url=url, headers=headers)
+        # Send cancel_payload, and receive response string, removing the encoding tag which causes issue with lxml parsing
+        response = client.service.handle_rtml(cancel).replace(
+            'encoding="ISO-8859-1"', ''
+        )
+        response_rtml = etree.fromstring(response)
+        mode = response_rtml.get('mode')
+        uid = response_rtml.get('uid')
+        if mode == 'confirm':
+
+            request.status = "deleted"
+
+            transaction = FacilityTransaction(
+                request=http.serialize_requests_request_xml(cancel),
+                response=http.serialize_requests_response_xml(response),
+                followup_request=request,
+                initiator_id=request.last_modified_by_id,
+            )
+            DBSession().add(transaction)
+
+
+class IOOAPI(LTAPI):
+
+    """An interface to LT IOO operations."""
+
+    # subclasses *must* implement the method below
+    @staticmethod
+    def submit(request):
+
+        """Submit a follow-up request to LT's IOO.
+
+        Parameters
+        ----------
+        request: skyportal.models.FollowupRequest
+            The request to add to the queue and the SkyPortal database.
+        """
+
+        from ..models import DBSession, FacilityTransaction
+
+        altdata = request.allocation.load_altdata()
+        if not altdata:
+            raise ValueError('Missing allocation information.')
+        ltreq = IOOIOIRequest()
+        observation_payload = ltreq._build_prolog()
+        ltreq._build_project(observation_payload, request)
+        ltreq._build_inst_schedule("IO:O", observation_payload, request)
+
+        headers = {
+            'Username': altdata["username"],
+            'Password': altdata["password"],
+        }
+        url = f"http://{cfg['app.lt_host']}:{cfg['app.lt_port']}/node_agent2/node_agent?wsdl"
+        client = Client(url=url, headers=headers)
+        full_payload = etree.tostring(
+            observation_payload, encoding="unicode", pretty_print=True
+        )
+        # Send payload, and receive response string, removing the encoding tag which causes issue with lxml parsing
+        response = client.service.handle_rtml(full_payload).replace(
+            'encoding="ISO-8859-1"', ''
+        )
+        response_rtml = etree.fromstring(response)
+        mode = response_rtml.get('mode')
+
+        if mode == 'confirm':
+            request.status = 'submitted'
+        else:
+            error = list(response_rtml.iter('{http://www.rtml.org/v3.1a}Error'))[0].text
```

Comment:
Is this operation safe ? I.e., will this token always precede the response error:?

Reply:
This has been the case for all of the errors I have seen. Maybe we put it in a try: except: and state error unknown if not?

Reply:
Does LT document this ?